src/main/java/com/rocketden/main/service/SubmitService.java

```diff
@@ -58,4 +105,58 @@ public SubmissionDto submitSolution(Game game, SubmissionRequest request) {
 
         return GameMapper.submissionToDto(submission);
     }
+
+    // Sends a POST request to the tester service to judge the user submission
+    protected Submission callTesterService(TesterRequest request) {
+        try {
+            HttpPost post = new HttpPost(getTesterUrl());
+
+            StringEntity stringEntity = new StringEntity(gson.toJson(request));
+            post.setEntity(stringEntity);
+            post.setHeader("Content-type", "application/json");
+
+            HttpResponse response = httpClient.execute(post);
+            String jsonResponse = EntityUtils.toString(response.getEntity());
+
+            TesterResponse testerResponse = getResponseFromJson(jsonResponse);
+
+            // TODO: logic to convert TesterResponse into Submission object
+            Submission submission = new Submission();
+            submission.setNumCorrect(request.getProblem().getTestCases().size());
+            submission.setNumTestCases(request.getProblem().getTestCases().size());
+            submission.setPlayerCode(new PlayerCode(request.getCode(), request.getLanguage()));
+
+            return submission;
+        } catch (Exception e) {
+            // If in debug mode (tester is unavailable), return a dummy submission
+            if (getDebugMode()) {
+                return getDummySubmission(request);
+            }
+
+            throw new ApiException(GameError.TESTER_ERROR);
+        }
+    }
+
+    // Is null in certain testing environments; if so, return a default value
+    private String getTesterUrl() {
+        if (testerUrl == null) {
+            return "http://localhost:8080";
```

Comment:
Will this be accurate, or is this just a default URL that may or may not connect to the tester service? If it's the latter, I may favor returning null and doing a check? I'm not sure -- what do you think.

Reply:
This is just returning a default URL (that doesn't work, since for our tests we won't have access to the tester service unless we link it with the live site which probably isn't good practice haha) for testing purposes - there's a few tests in which it can't read in the value for testerUrl, so this is just checking that and setting it to the same value as in the test application.properties

About the null check - could you clarify what that might look like?

Reply:
I guess not actually returning null, but maybe throwing an error as the application has no chance of working in those testing environments (if I understand correctly). I don't have a huge preference on this though.

Reply:
I'll probably leave it as is then - it's a little gimmicky but I think those unit tests are still valid/useful to have